### PR TITLE
Ml constant variable placeholder implementation

### PR DESCRIPTION
**Pull request by baykaner**
- implements abstract DataHolder, from which Constant, Variable and Placeholder derive.
These three have different characteristics as follows:

Constant, Variable, Placeholder
- Trainable: No, Yes, No
- Mutable: No, Yes, Yes
- Shareable: Yes, Yes, No
- Serializes data: Yes, Yes, No

### CODE REVIEW

File: libs/ml/include/ml/ops/constant.hpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+namespace fetch {
+namespace ml {
+namespace ops {
+
+/**
+ * A Constant is a DataHolder intended to store an immutable value.
+ * It has the following features:
+ * 1. trainable: no
+ * 2. mutable: no, the data can be only written once
+ * 3. shareable: yes, shared layers can re-used constants
```
Comment (evsmithx) [inline libs/ml/include/ml/ops/constant.hpp]: used -> use

File: libs/ml/include/ml/ops/constant.hpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+  explicit Constant(SPType const &sp)
+    : DataHolder<T>(sp)
+  {}
+
+  ~Constant() override = default;
+
+  std::shared_ptr<OpsSaveableParams> GetOpSaveableParams() override
+  {
+    return std::make_shared<SPType>();
```
Comment (evsmithx) [inline libs/ml/include/ml/ops/constant.hpp]: Shouldn't Constant store its data_ here?

File: libs/ml/include/ml/saveparams/saveable_params.hpp [hunk 3]
```diff
@@ -491,8 +494,21 @@ struct LayerMultiHeadSaveableParams : public SubGraphSaveableParams<TensorType>
...
+struct OpDataHolderSaveableParams : public OpsSaveableParams
+{
+  fetch::ml::OpType           op_type = OpType::OP_DATAHOLDER;
+  std::shared_ptr<TensorType> data;
+};
+
+template <typename TensorType>
+struct OpConstantSaveableParams : public OpDataHolderSaveableParams<TensorType>
+{
+  fetch::ml::OpType op_type = OpType::OP_CONSTANT;
```
Comment (evsmithx) [inline libs/ml/include/ml/saveparams/saveable_params.hpp:507]: Constant should also store data

File: libs/ml/include/ml/utilities/graph_builder.hpp [hunk 4]
```diff
@@ -160,6 +160,13 @@ void BuildNodeAndInsertTrainables(NodeSaveableParams<T> const &nsp, std::string
     g->AddTrainable(node, name);
     break;
   }
+  case ops::Constant<T>::OpCode():
+  {
+    op_ptr = GetOp<ops::Constant<T>>(nsp.op_save_params);
+    node->SetNodeSaveableParams(nsp, op_ptr);
+    g->AddTrainable(node, name);
+    break;
+  }
```
Comment (evsmithx) [inline libs/ml/include/ml/utilities/graph_builder.hpp:169]: Do Variable and Placeholder also need to be added here?

**Review by evsmithx — CHANGES_REQUESTED** [review]
There's a hack in the test for scaled dot product attention that should be removed now that we are solving the problem properly (there's a todo next to it). Also there should be tests for the new ops.

File: libs/ml/include/ml/utilities/graph_builder.hpp (hunk 4 above)
Comment (baykaner) [reply libs/ml/include/ml/utilities/graph_builder.hpp:169]: well... not for variable, no, because Variable is an abstract class. but Placeholder, yes

Comment (baykaner) [reply libs/ml/include/ml/utilities/graph_builder.hpp:169]: but placeholder was already there of course

File: libs/ml/tests/ml/layers/scaled_dot_product_attention.cpp [hunk 5]
```diff
@@ -339,10 +339,10 @@ TYPED_TEST(ScaledDotProductAttention, saveparams_test)
...
   layer2.SetInput("ScaledDotProductAttention_Mask", mask_data);
 
-  // todo(#1628): temporary fix because placeholders are discarded in serialisation
-  TypeParam sqrt_dk_tensor = std::vector<SizeType>({1, 1, 1});
-  sqrt_dk_tensor(0, 0, 0)  = fetch::math::Sqrt(static_cast<DataType>(key_dim));
-  layer2.SetInput("ScaledDotProductAttention_Sqrt_Key_Dim", sqrt_dk_tensor);
+  //  // todo(#1628): temporary fix because placeholders are discarded in serialisation
+  //  TypeParam sqrt_dk_tensor = std::vector<SizeType>({1, 1, 1});
+  //  sqrt_dk_tensor(0, 0, 0)  = fetch::math::Sqrt(static_cast<DataType>(key_dim));
+  //  layer2.SetInput("ScaledDotProductAttention_Sqrt_Key_Dim", sqrt_dk_tensor);
```
Comment (evsmithx) [inline libs/ml/tests/ml/layers/scaled_dot_product_attention.cpp]: Commented-out code

**Review by evsmithx — DISMISSED** [review]
Some commented code

**Review by evsmithx — APPROVED** [review]
LGTM

**Pull request merged**